examples/chatgpt/gpt_actions_library/gpt_action_mongodb_vector.md

```diff
@@ -0,0 +1,356 @@
+async function getEmbedding(searchTerm) {
+  try {
+    const embeddingResponse = await openai.embeddings.create({
+      model: process.env.OPENAI_MODEL, // Specify the embedding model from env
+      input: searchTerm,
+    });
+
+    const embedding = embeddingResponse.data[0].embedding;
+    return embedding;
+  } catch (error) {
+    console.error("Error getting embedding from OpenAI:", error);
+    throw error;
+  }
+}
+
+// Main function to handle the HTTP request and authenticate users
+functions.http('queryVectorDB', async (req, res) => {
+  const authHeader = req.headers.authorization;
+
+  if (!authHeader) {
+    return res.status(401).send('Unauthorized: No token provided');
+  }
+
+  const token = authHeader.split(' ')[1];
+  if (!token) {
+    return res.status(401).send('Unauthorized: No token provided');
+  }
+
+  try {
+    // Validate the access token
+    await validateAccessToken(token);
+
+    // Proceed with your logic after successful authentication
+    const searchTerm = req.body.searchTerm || "";
+    if (typeof searchTerm !== 'string' || searchTerm.trim().length === 0) {
+      return res.status(400).send("Invalid searchTerm: it must be a non-empty string.");
+    }
+
+    // Get the query vector from OpenAI
+    const queryVector = await getEmbedding(searchTerm);
+
+    // Define the aggregation pipeline for vector search - switch this out to match your collection
+    const aggPipeline = [
+      {
+        '$vectorSearch': {
+          'index': process.env.MONGODB_VECTOR_INDEX, // Your vector index name from env
+          'path': 'plot_embedding', // Path to the vector field in your documents
+          'queryVector': queryVector, // The vector from OpenAI
+          'numCandidates': 150, // Number of candidates to consider
+          'limit': 10 // Limit the number of results returned
+        }
+      },
+      {
+        '$project': {
+          '_id': 0,
+          'plot': 1,
+          'title': 1,
+          'score': {
+            '$meta': 'vectorSearchScore' // Include the vector search score
+          }
+        }
+      }
+    ];
+
+    // Perform the search using MongoDB
+    const results = await searchMongoDB(aggPipeline);
+    res.status(200).json(results);
+
+  } catch (error) {
+    console.error("Authentication or search error:", error);
+    res.status(401).send('Unauthorized: Invalid token or internal error');
+  }
+});
+
+```
+
+Make sure you have included these dependencies
+
+```
+  "dependencies": {
+    "@google-cloud/functions-framework": "^3.4.2",
+    "axios": "^1.7.7",
+    "dotenv": "^16.4.5",
+    "mongodb": "^6.8.0",
+    "openai": "^4.62.1"
+  }
+```
+Your `.env` file should contain the following values
+
+```
+OPENAI_API_KEY=
+MONGODB_URI=
+MONGODB_DB_NAME=
+MONGODB_COLLECTION_NAME=
+OPENAI_MODEL=
+MONGODB_VECTOR_INDEX=
+```
+
+## Application Setup
+
+### Create vector index
+
+Ensure that your MongoDB collection has a vector index. Here is an example of how to create one
```

Comment:
Do we have a script to run this on the MONGODB_URI or a command line to connect to it?